File: 2025-01-January-GFG-POTD/January 02 - Subarrays with sum K/Solution.py
from collections import defaultdict

class Solution:
    def countSubarrays(self, arr, k):
        prefix_sum_count = defaultdict(int)
        current_sum = 0
        count = 0

        # Initialize the prefix sum count with 0 sum
        prefix_sum_count[0] = 1

        for num in arr:
            current_sum += num

            # Check if there exists a prefix sum such that currentSum - prefixSum = k
            if current_sum - k in prefix_sum_count:
                count += prefix_sum_count[current_sum - k]

            # Update the frequency of the current prefix sum in the map
            prefix_sum_count[current_sum] += 1

        return count

if __name__ == "__main__":
    t = int(input())

    for _ in range(t):
        k = int(input())
        arr = list(map(int, input().split()))

        sol = Solution()
        res = sol.countSubarrays(arr, k)

        print(res)
        print("~")

File: 2025-01-January-GFG-POTD/January 02 - Subarrays with sum K/Solution.java
//{ Driver Code Starts
// Initial Template for Java

import java.io.*;
import java.lang.*;
import java.util.*;

public class Main {
    public static void main(String args[]) throws IOException {
        // taking input using class Scanner
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(br.readLine());

        while (t-- > 0) {
            // taking total number of elements
            int k = Integer.parseInt(br.readLine());
            String line = br.readLine();
            String[] tokens = line.split(" ");

            // Create an ArrayList to store the integers
            ArrayList<Integer> array = new ArrayList<>();

            // Parse the tokens into integers and add to the array
            for (String token : tokens) {
                array.add(Integer.parseInt(token));
            }

            int[] arr = new int[array.size()];
            int idx = 0;
            for (int i : array) arr[idx++] = i;
            int res = new Solution().countSubarrays(arr, k);

            System.out.print(res);
            System.out.println();
            System.out.println("~");
        }
    }
}
// } Driver Code Ends


// User function Template for Java
class Solution {
    public int countSubarrays(int arr[], int k) {
        // code here
        // HashMap to store prefix sums and their frequencies
        HashMap<Integer, Integer> prefixSumCount = new HashMap<>();
        int currentSum = 0;
        int count = 0;

        // Initialize the prefix sum count with 0 sum (to handle cases where a subarray starts from index 0)
        prefixSumCount.put(0, 1);

        for (int num : arr) {
            currentSum += num;

            // Check if there exists a prefix sum such that currentSum - prefixSum = k
            if (prefixSumCount.containsKey(currentSum - k)) {
                count += prefixSumCount.get(currentSum - k);
            }

            // Update the frequency of the current prefix sum in the map
            prefixSumCount.put(currentSum, prefixSumCount.getOrDefault(currentSum, 0) + 1);
        }

        return count;
    }
}